pkg/asset/imagebased/image/ignition.go

```diff
@@ -0,0 +1,183 @@
+
+	ibiConfig := configAsset.Config
+	if ibiConfig == nil {
+		return errors.New("imagebased-installation-config.yaml is required")
+	}
+
+	config := &igntypes.Config{
+		Ignition: igntypes.Ignition{
+			Version: igntypes.MaxVersion.String(),
+		},
+		Passwd: igntypes.Passwd{
+			Users: []igntypes.PasswdUser{
+				{
+					Name: "core",
+					SSHAuthorizedKeys: []igntypes.SSHAuthorizedKey{
+						igntypes.SSHAuthorizedKey(ibiConfig.SSHKey),
+					},
+				},
+			},
+		},
+	}
+
+	setFileInIgnition(config, pullSecretPath, ibiConfig.PullSecret, 0o600)
+
+	ibiConfigFile := ibiConfigurationFile{
+		ExtraPartitionLabel:  ibiConfig.ExtraPartitionLabel,
+		ExtraPartitionNumber: ibiConfig.ExtraPartitionNumber,
+		ExtraPartitionStart:  ibiConfig.ExtraPartitionStart,
+		InstallationDisk:     ibiConfig.InstallationDisk,
+		PrecacheDisabled:     ibiConfig.PrecacheDisabled,
+		PrecacheBestEffort:   ibiConfig.PrecacheBestEffort,
+		SeedVersion:          ibiConfig.SeedVersion,
+		SeedImage:            ibiConfig.SeedImage,
+		Shutdown:             ibiConfig.Shutdown,
+		SkipDiskCleanup:      ibiConfig.SkipDiskCleanup,
+		UseContainersFolder:  ibiConfig.UseContainersFolder,
+	}
+	marshaled, err := json.Marshal(ibiConfigFile)
+	if err != nil {
+		return fmt.Errorf("failed to marshall image-based installation configuration data: %w", err)
+	}
+	setFileInIgnition(config, ibiConfigurationPath, string(marshaled), 0o600)
+
+	if len(registriesConf.Data) > 0 {
+		setFileInIgnition(config, registriesConfPath, string(registriesConf.Data), 0o600)
+	}
+
+	setFileInIgnition(config, installationScriptPath, installationScript, 0o755)
+	setUnitInIgnition(config, installationServiceName, installationServiceAsset.Content)
+
+	if ibiConfig.AdditionalTrustBundle != "" {
+		setFileInIgnition(config, trustedBundlePath, ibiConfig.AdditionalTrustBundle, 0o600)
+	}
+
+	if ibiConfig.NetworkConfig.String() != "" {
+		setFileInIgnition(config, nmstateConfigPath, ibiConfig.NetworkConfig.String(), 0o600)
+		setUnitInIgnition(config, networkConfigServiceName, networkConfigService)
+	}
+
+	if ibiConfig.IgnitionConfigOverride != "" {
+		ignitionConfigOverride, _, err := v3_2.Parse([]byte(ibiConfig.IgnitionConfigOverride))
+		if err != nil {
+			return fmt.Errorf("failed to parse ignition config override: %w", err)
+		}
+
+		merged, _ := merge.MergeStructTranscribe(*config, ignitionConfigOverride)
+		marshaledMerged, err := json.Marshal(merged)
+		if err != nil {
+			return fmt.Errorf("failed to marshal merged ignition config: %w", err)
+		}
+
+		if err := json.Unmarshal(marshaledMerged, config); err != nil {
+			return fmt.Errorf("failed to unmarshal merged ignition config: %w", err)
+		}
+	}
+
+	if postDeployment.File != nil {
+		setFileInIgnition(config, postDeploymentScriptPath, string(postDeployment.File.Data), 0o755)
+	}
+
+	i.Config = config
+	return nil
+}
+
+func setFileInIgnition(config *igntypes.Config, filePath, fileContents string, mode int) {
```

Comment:
pretty sure this is already implemented elsewhere

Reply:
Thanks for the suggestion, I remembered I had found [these](https://github.com/openshift/installer/blob/master/pkg/asset/ignition/node.go#L37-L59) and [these](https://github.com/openshift/installer/blob/master/pkg/asset/ignition/bootstrap/common.go#L358-L525) when looking into the respective agent code:

- https://github.com/openshift/installer/blob/master/pkg/asset/ignition/node.go#L37-L59
- https://github.com/openshift/installer/blob/master/pkg/asset/agent/image/ignition.go#L301-L304

but I thought they were more complicated than what we needed and decided to add them. I'll check how this works with the aforementioned/existing functions and update the PR, we could indeed skip adding those helper functions :+1:

Reply:
@eranco74 I've updated the ignition file and systemd units addition to leverage the https://github.com/openshift/installer/tree/master/pkg/asset/ignition pkg and removed the previously defined helpers. It now blends well with the rest of the installer codebase, it should be cleaner and more readable now. PTAL